### PR TITLE
Rewriting encryption/decryption back to Sodium v1

### PR DESCRIPTION
**Pull request by makingthematrix**
After encrypting a backup, decryption fails.  For now I think that the problem is with a nonce (which we don't use and maybe we should). But that's only a theory. If you want to help, please checkout this PR, build it, install the apk on your phone, try encrypting and decrypting, and look through the code while you do it. Maybe you will find something. 
I made an extension method `ByteArray.describe` that displays a byte array in a human-friendly way.

Also, read this:
https://libsodium.gitbook.io/doc/secret-key_cryptography/aead/chacha20-poly1305/xchacha20-poly1305_construction
#### APK
[Download build #2705](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2705/artifact/build/artifact/wire-dev-PR3007-2705.apk)
[Download build #2710](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2710/artifact/build/artifact/wire-dev-PR3007-2710.apk)

### CODE REVIEW

File: app/src/main/kotlin/com/waz/zclient/core/extension/String.kt [hunk 1]
```diff
@@ -11,3 +11,14 @@ fun String.toSpanned() =
     } else {
         Html.fromHtml(this)
     }
+
+fun ByteArray.describe(splitAt: Int = 4): String {
```
Comment (BradleyWire) [inline app/src/main/kotlin/com/waz/zclient/core/extension/String.kt:15]: Would be good to unit test this

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt [hunk 2]
```diff
@@ -29,32 +32,41 @@ class DecryptionHandler(
...
-                val decrypted = ByteArray(cipherText.size + crypto.aBytesLength())
-                when (crypto.generatePullMessagePart(state, decrypted, cipherText)) {
-                    0 -> Either.Right(decrypted)
-                    else -> Either.Left(DecryptionFailed)
-                }
-            }
+    private fun decrypt(cipherText: ByteArray, key: ByteArray): Either<Failure, ByteArray> {
+        val decrypted = ByteArray(cipherText.size - crypto.aBytesLength())
+        verbose(TAG, "CRY decrypt, cipherText: ${cipherText.describe()}")
+        return when (crypto.decrypt(decrypted, cipherText, key)) {
```
Comment (BradleyWire) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt:61]: Does this not break unit tests?

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/header/EncryptionHeader.kt [hunk 3]
```diff
@@ -44,21 +46,24 @@ class EncryptionHeaderMapper {
 
     fun toByteArray(header: EncryptedBackupHeader): ByteArray =
         ByteBuffer.allocate(TOTAL_HEADER_LENGTH).apply {
-            put(ANDROID_MAGIC_NUMBER.toByteArray())
+            Logger.verbose(TAG, "android magic number: ${ANDROID_MAGIC_NUMBER.contentToString()}")
```
Comment (BradleyWire) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/header/EncryptionHeader.kt:49]: Is this log needed?

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/Crypto.kt [hunk 4]
```diff
@@ -24,34 +24,6 @@ class Crypto(private val cryptoWrapper: CryptoWrapper) {
         }
     }
 
-    internal fun initEncryptState(initKey: ByteArray, initHeader: ByteArray) =
```
Comment (BradleyWire) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/Crypto.kt:27]: You need to remove the test cases for these changes

**Review by BradleyWire — CHANGES_REQUESTED** [review]
Unit testing is lacking in this PR, I don't believe they all are passing in its current state

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt (hunk 2 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt]: Yes. I fixed the tests.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/Crypto.kt (hunk 4 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/Crypto.kt:27]: done

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/header/EncryptionHeader.kt (hunk 3 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/header/EncryptionHeader.kt]: No. Removed.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/CryptoWrapper.kt [hunk 5]
```diff
@@ -6,40 +6,42 @@ import org.libsodium.jni.Sodium
...
+        Sodium.crypto_aead_xchacha20poly1305_ietf_npubbytes()
 
-    fun generatePushMessagePart(state: ByteArray, cipherText: ByteArray, msg: ByteArray) =
-        Sodium.crypto_secretstream_xchacha20poly1305_push(
-            state,
+    fun encrypt(cipherText: ByteArray, msg: ByteArray, key: ByteArray, nonce: ByteArray): Int =
+        Sodium.crypto_aead_xchacha20poly1305_ietf_encrypt(
             cipherText,
-            intArrayOf(),
+            intArrayOf(1),
```
Comment (franziskuskiefer) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/CryptoWrapper.kt:24]: This creates a signed 32-bit integer array. But we need a pointer to an unsigned 64-bit integer.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/CryptoWrapper.kt [hunk 6]
```diff
@@ -6,40 +6,42 @@ import org.libsodium.jni.Sodium
...
             msg,
             msg.size,
             byteArrayOf(),
             0,
-            Sodium.crypto_secretstream_xchacha20poly1305_tag_final().toShort()
+            byteArrayOf(),
```
Comment (franziskuskiefer) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/CryptoWrapper.kt:29]: We should add some associated data here... But let's skip for now.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt [hunk 7]
```diff
@@ -21,39 +21,41 @@ class DecryptionHandler(
...
         }
     }
 
-    private fun decryptBackupFile(password: String, backupFile: File, salt: ByteArray): Either<Failure, File> {
-        val encryptedBackupBytes = ByteArray(TOTAL_HEADER_LENGTH)
-        backupFile.inputStream().buffered().read(encryptedBackupBytes)
-        return decryptWithHash(encryptedBackupBytes, password, salt).map { decryptedBackupBytes ->
-            File.createTempFile("wire_backup", ".zip").apply { writeBytes(decryptedBackupBytes) }
+    private fun decryptBackupFile(password: String, backupFile: File, salt: ByteArray, nonce: ByteArray): Either<Failure, File> {
+        val backupLength = backupFile.length() - TOTAL_HEADER_LENGTH
```
Comment (franziskuskiefer) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt:32]: What happens if `backupFile.length() < TOTAL_HEADER_LENGTH`?

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt [hunk 8]
```diff
@@ -21,39 +21,41 @@ class DecryptionHandler(
...
-            .flatMap { state ->
-                val cipherText = input.drop(crypto.streamHeaderLength()).toByteArray()
-                val decrypted = ByteArray(cipherText.size + crypto.aBytesLength())
-                when (crypto.generatePullMessagePart(state, decrypted, cipherText)) {
-                    0 -> Either.Right(decrypted)
-                    else -> Either.Left(DecryptionFailed)
-                }
-            }
+    private fun decrypt(cipherText: ByteArray, key: ByteArray, nonce: ByteArray): Either<Failure, ByteArray> {
+        val decrypted = ByteArray(cipherText.size - crypto.aBytesLength())
```
Comment (franziskuskiefer) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt:54]: What if `cipherText.size < crypto.aBytesLength()`?

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/header/EncryptionHeader.kt [hunk 9]
```diff
@@ -1,21 +1,23 @@
...
 private const val SALT_LENGTH = 16
+private const val NONCE_LENGTH = 24
 private const val ANDROID_MAGIC_NUMBER_LENGTH = 4
 private const val TAG = "EncryptionHeader"
-internal const val CURRENT_VERSION: Short = 2
+internal const val CURRENT_VERSION: Short = 1
 
 const val UUID_HASH_LENGTH = 32
-const val TOTAL_HEADER_LENGTH = ANDROID_MAGIC_NUMBER_LENGTH + 1 + 2 + SALT_LENGTH + UUID_HASH_LENGTH + 4 + 4
+const val TOTAL_HEADER_LENGTH = ANDROID_MAGIC_NUMBER_LENGTH + 1 + 2 + SALT_LENGTH + UUID_HASH_LENGTH + 4 + 4 + NONCE_LENGTH
```
Comment (franziskuskiefer) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/header/EncryptionHeader.kt:20]: So I figured out why this isn't in the spec. iOS implements v1 and appears to actually use the streaming API, though that's not actually specified in v1. Let's go with this. All this needs to be cleaned up at some point though.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Generally looks good.
A couple places need tests and there's one issue in encrypt and decrypt that needs fixing.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/CryptoWrapper.kt (hunk 5 above)
Comment (franziskuskiefer) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/CryptoWrapper.kt:24]: Ok, then this appears to be an issue in the JNI wrapper of libsodium.

Comment (franziskuskiefer) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/CryptoWrapper.kt:24]: What JNI bindings do we use here? This should be filed to get it fixed.

File: app/src/main/kotlin/com/waz/zclient/core/extension/String.kt (hunk 1 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/core/extension/String.kt:15]: ok

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/CryptoWrapper.kt (hunk 5 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/CryptoWrapper.kt:24]: @franziskuskiefer : `com.github.joshjdevl.libsodiumjni:libsodium-jni-aar:2.0.2`

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt (hunk 7 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt]: Right. I changed this code a bit and now it returns `Either.Left(DecryptionFailed)` if the file is too short. I also added a try/catch block for reading from the file.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt (hunk 8 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt]: Changed similar as above.

File: app/src/main/kotlin/com/waz/zclient/core/extension/String.kt [hunk 10]
```diff
@@ -17,7 +17,7 @@ fun ByteArray.describe(splitAt: Int = 4): String {
         if (size <= splitAt * 2) contentToString()
         else {
             val head = take(splitAt).map { it.toInt() }.joinToString(", ")
-            val tail = drop(size - splitAt + 1).map { it.toInt() }.joinToString(", ")
+            val tail = drop(size - splitAt).map { it.toInt() }.joinToString(", ")
```
Comment (franziskuskiefer) [inline app/src/main/kotlin/com/waz/zclient/core/extension/String.kt:20]: Did you find this with the test you added?

**Review by franziskuskiefer — APPROVED** [review]
Thanks. lgtm now.
I filed https://github.com/joshjdevl/libsodium-jni/issues/142 to see if we can get the JNI issue fixed. It's nothing critical (we'll run out of memory before we run into an issue here ;).

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/Crypto.kt [hunk 11]
```diff
@@ -24,90 +21,57 @@ class Crypto(private val cryptoWrapper: CryptoWrapper) {
...
+    }
+
     internal fun hashWithMessagePart(input: String, salt: ByteArray): Either<Failure, ByteArray> {
         val output = ByteArray(encryptExpectedKeyBytes())
         val passBytes = input.toByteArray()
-        val pushMessage = generatePwhashMessagePart(output, passBytes, salt)
-        return pushMessage.takeIf { it == 0 }?.let { Either.Right(output) }
-            ?: Either.Left(HashingFailed)
+        val pushMessage = cryptoWrapper.generatePwhashMessagePart(output, passBytes, salt)
+        return pushMessage.takeIf { it == 0 }?.let { Either.Right(output) } ?: Either.Left(HashingFailed)
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/Crypto.kt:42]: i'd prefer if/else here

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/header/EncryptionHeader.kt (hunk 9 above)
Comment (gizemb) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/header/EncryptionHeader.kt:20]: should we put TODOs for such things?

File: app/src/test/kotlin/com/waz/zclient/core/extension/StringTest.kt [hunk 12]
```diff
@@ -0,0 +1,33 @@
+package com.waz.zclient.core.extension
+
+import com.waz.zclient.UnitTest
+import org.amshove.kluent.shouldContain
+import org.amshove.kluent.shouldEqual
+import org.junit.Test
+
+class StringTest : UnitTest() {
```
Comment (gizemb) [inline app/src/test/kotlin/com/waz/zclient/core/extension/StringTest.kt:8]: ByteArrayExtensionsTest

File: app/src/test/kotlin/com/waz/zclient/feature/backup/crypto/CryptoTest.kt [hunk 13]
```diff
@@ -275,21 +145,10 @@ class CryptoTest : UnitTest() {
...
-
-        `when`(cryptoWrapper.secretStreamPolyKeyBytes()).thenReturn(bytes)
-
-        val pullBytes = crypto.decryptExpectedKeyBytes()
-
-        assertEquals(pullBytes, bytes)
-    }
-
     @Test
     fun `given hash size is same as expected size, then propagate success`() {
```
Comment (gizemb) [inline app/src/test/kotlin/com/waz/zclient/feature/backup/crypto/CryptoTest.kt:149]: "when `checkExpectedKeySize` is called, "

File: app/src/test/kotlin/com/waz/zclient/feature/backup/crypto/CryptoTest.kt [hunk 14]
```diff
@@ -298,8 +157,8 @@ class CryptoTest : UnitTest() {
 
     @Test
     fun `given hash size is not the same as expected size, then propagate HashWrongSize error`() {
```
Comment (gizemb) [inline app/src/test/kotlin/com/waz/zclient/feature/backup/crypto/CryptoTest.kt:159]: same here

File: app/src/test/kotlin/com/waz/zclient/feature/backup/crypto/encryption/EncryptionHandlerTest.kt [hunk 15]
```diff
@@ -32,57 +39,119 @@ class EncryptionHandlerTest : UnitTest() {
...
+        `when`(crypto.generateSalt()).thenReturn(Either.Right(salt))
+        `when`(crypto.generateNonce()).thenReturn(Either.Right(nonce))
+        `when`(crypto.hashWithMessagePart(any(), any())).thenReturn(Either.Right(hash))
+        `when`(headerMetaData.createMetaData(any(), any(), any())).thenReturn(Either.Right(byteArrayOf()))
+        `when`(crypto.encryptExpectedKeyBytes()).thenReturn(ENCRYPTION_HASH_BYTES)
+        `when`(crypto.checkExpectedKeySize(wrongHashBytesSize, ENCRYPTION_HASH_BYTES)).thenReturn(Either.Left(HashWrongSize))
 
+        encryptionHandler.encryptBackup(backupFile, userId, password, backupFile.name + "_encrypted")
+            .onSuccess { fail("If the key size check fails, the encryption should fail") }
+            .onFailure { it shouldEqual HashWrongSize }
```
Comment (gizemb) [inline app/src/test/kotlin/com/waz/zclient/feature/backup/crypto/encryption/EncryptionHandlerTest.kt:151]: we can use assertLeft/assertRight now for convenience. Bradley added them.

File: app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCaseTest.kt [hunk 16]
```diff
@@ -65,8 +65,7 @@ class CreateBackUpUseCaseTest : UnitTest() {
             verify(repo3).saveBackup()
             verify(metaDataHandler).generateMetaDataFile(metaData)
             verify(zipHandler).zip(anyString(), anyList())
-            // TODO: Uncomment when the encryption is ready
-            // verify(encryptionHandler).encrypt(any(), any(), anyString())
+            verify(encryptionHandler).encryptBackup(any(), any(), anyString(), anyString())
 
             assert(result.isRight)
```
Comment (gizemb) [inline app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCaseTest.kt:70]: it should be changed to junit's `assertEquals`. could you also check if there are any other usages of this `assert`?

**Review by gizemb — CHANGES_REQUESTED** [review]
I think structurally it's ok, but I request changes for the wrong `assert` usage in unit tests.

File: app/src/main/kotlin/com/waz/zclient/core/extension/String.kt (hunk 10 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/core/extension/String.kt:20]: @franziskuskiefer : It's not a bug. It's just a matter of what we want to see in the log statements. I thought that the `splitAt` argument may be understood as "I want this many bytes visible in the front and as well at the back". So now for `splitAt=4` the result will be `[1, 2, 3, 4, ..., 11, 12, 13, 14]`. Previously, `11` was be invisible.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt [hunk 17]
```diff
@@ -2,59 +2,81 @@ package com.waz.zclient.feature.backup.crypto.decryption
...
-    private fun decryptBackupFile(password: String, backupFile: File, salt: ByteArray): Either<Failure, File> {
-        val encryptedBackupBytes = ByteArray(TOTAL_HEADER_LENGTH)
-        backupFile.inputStream().buffered().read(encryptedBackupBytes)
-        return decryptWithHash(encryptedBackupBytes, password, salt).map { decryptedBackupBytes ->
-            File.createTempFile("wire_backup", ".zip").apply { writeBytes(decryptedBackupBytes) }
+    private fun decryptBackupFile(password: String, backupFile: File, salt: ByteArray, nonce: ByteArray): Either<Failure, File> =
+        readCipherText(backupFile).flatMap { cipherText ->
+            decryptWithHash(cipherText, password, salt, nonce)
+        }.map { decryptedBackupBytes ->
+            File.createTempFile("wire_backup", ".zip").apply {
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/crypto/decryption/DecryptionHandler.kt:38]: Avoid hardcoding. Create constants inside the companion object.

```
const val TMP_FILE_NAME = "wire_backup"
const val TMP_FILE_EXTENSION = ".zip"
```

File: app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/RestoreBackUpUseCaseTest.kt [hunk 18]
```diff
@@ -35,7 +36,7 @@ class RestoreBackUpUseCaseTest : UnitTest() {
     private val userId = UserId.apply(UUID.randomUUID().toString())
     private val password = "password"
     private val metadataFile = File(MetaDataHandler.FILE_NAME) // a mock file, don't create it
-    private val zipFile = File("mocked.zip")
+    private val zipFile = File("file_zipped.zip")
```
Comment (mejdoo) [inline app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/RestoreBackUpUseCaseTest.kt:39]: same for the tests

**Review by mejdoo — CHANGES_REQUESTED** [review]

File: app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCaseTest.kt (hunk 16 above)
Comment (makingthematrix) [reply app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCaseTest.kt:70]: @gizemb : Quite a lot in various unit tests. I don't want to change them all now. I just fixed one more in the backups unit tests.

Comment (gizemb) [reply app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCaseTest.kt]: @makingthematrix that's fair, no problem :)

**Review by mejdoo — APPROVED** [review]

**Review by gizemb — APPROVED** [review]

**Pull request merged**